jerry-core/ecma/base/ecma-gc.c

```diff
@@ -399,14 +426,41 @@ ecma_gc_sweep (ecma_object_t *object_p) /**< object to free */
   if (!ecma_is_lexical_environment (object_p)
       || ecma_get_lex_env_type (object_p) == ECMA_LEXICAL_ENVIRONMENT_DECLARATIVE)
   {
-    for (ecma_property_t *property = ecma_get_property_list (object_p), *next_property_p;
-         property != NULL;
-         property = next_property_p)
+    ecma_property_header_t *prop_iter_p = ecma_get_property_list (object_p);
+
+    while (prop_iter_p != NULL)
     {
-      next_property_p = ECMA_GET_POINTER (ecma_property_t,
-                                          property->next_property_p);
+      JERRY_ASSERT (ECMA_PROPERTY_IS_PROPERTY_PAIR (prop_iter_p));
+
+      /* Both cannot be deleted. */
+      JERRY_ASSERT (prop_iter_p->types[0].type_and_flags != ECMA_PROPERTY_TYPE_DELETED
+                    || prop_iter_p->types[1].type_and_flags != ECMA_PROPERTY_TYPE_DELETED);
+
+      ecma_property_pair_t *prop_pair_p = (ecma_property_pair_t *) prop_iter_p;
+
+      for (int i = 0; i < 2; i++)
+      {
+        if (prop_iter_p->types[i].type_and_flags != ECMA_PROPERTY_TYPE_DELETED)
+        {
+          ecma_string_t *name_p = ECMA_GET_POINTER (ecma_string_t, prop_pair_p->names_cp[i]);
+
+          ecma_free_property (object_p, name_p, prop_iter_p->types + i);
+
+          if (name_p != NULL)
+          {
+            ecma_deref_ecma_string (name_p);
+          }
+        }
+      }
+
+      /* Both must be deleted. */
+      JERRY_ASSERT (prop_iter_p->types[0].type_and_flags == ECMA_PROPERTY_TYPE_DELETED
+                    && prop_iter_p->types[1].type_and_flags == ECMA_PROPERTY_TYPE_DELETED);
```

Comment:
This assert can be moved into the loop above chacking `prop_iter_p->types[i].type_and_flags == ECMA_PROPERTY_TYPE_DELETED`.